Failing test with pandas 0.23

I am having a unit test error that is occurring with pandas 0.23 but not pandas 0.22:

```
=================================== FAILURES ===================================
___________________________ test_sanitize_dataframe ____________________________

    def test_sanitize_dataframe():
        # create a dataframe with various types
        df = pd.DataFrame({'s': list('abcde'),
                           'f': np.arange(5, dtype=float),
                           'i': np.arange(5, dtype=int),
                           'd': pd.date_range('2012-01-01', periods=5, freq='H'),
                           'c': pd.Series(list('ababc'), dtype='category')})
    
        # add some nulls
        df.ix[0, 's'] = None
        df.ix[0, 'f'] = np.nan
        df.ix[0, 'd'] = pd.NaT
    
        # JSON serialize. This will fail on non-sanitized dataframes
        df_clean = sanitize_dataframe(df)
        s = json.dumps(df_clean.to_dict(orient='records'))
    
        # Re-construct pandas dataframe
        df2 = pd.read_json(s)
    
        # Re-apply original types
        for col in df:
            if str(df[col].dtype).startswith('datetime'):
                # astype(datetime) introduces time-zone issues:
                # to_datetime() does not.
                df2[col] = pd.to_datetime(df2[col])
            else:
                df2[col] = df2[col].astype(df[col].dtype)
    
>       assert df.equals(df2)
E       assert False
E        +  where False = <bound method NDFrame.equals of       s    f  i                   d  c\n0  None  NaN  0                 NaT  a\n1     b ...b\n2     c  2.0  2 2012-01-01 02:00:00  a\n3     d  3.0  3 2012-01-01 03:00:00  b\n4     e  4.0  4 2012-01-01 04:00:00  c>(   c                   d    f  i     s\n0  a                 NaT  NaN  0  None\n1  b 2012-01-01 01:00:00  1.0  1     b\n2  a 2012-01-01 02:00:00  2.0  2     c\n3  b 2012-01-01 03:00:00  3.0  3     d\n4  c 2012-01-01 04:00:00  4.0  4     e)
E        +    where <bound method NDFrame.equals of       s    f  i                   d  c\n0  None  NaN  0                 NaT  a\n1     b ...b\n2     c  2.0  2 2012-01-01 02:00:00  a\n3     d  3.0  3 2012-01-01 03:00:00  b\n4     e  4.0  4 2012-01-01 04:00:00  c> =       s    f  i                   d  c\n0  None  NaN  0                 NaT  a\n1     b  1.0  1 2012-01-01 01:00:00  b\n2     c  2.0  2 2012-01-01 02:00:00  a\n3     d  3.0  3 2012-01-01 03:00:00  b\n4     e  4.0  4 2012-01-01 04:00:00  c.equals

vega/tests/test_utils.py:59: AssertionError
```

--- a/vega/tests/test_utils.py
+++ b/vega/tests/test_utils.py
@@ -32,13 +32,16 @@ def test_sanitize_dataframe():
     df = pd.DataFrame({'s': list('abcde'),
                        'f': np.arange(5, dtype=float),
                        'i': np.arange(5, dtype=int),
+                       'b': np.array([True, False, True, True, False]),
                        'd': pd.date_range('2012-01-01', periods=5, freq='H'),
-                       'c': pd.Series(list('ababc'), dtype='category')})
+                       'c': pd.Series(list('ababc'), dtype='category'),
+                       'o': pd.Series([np.array(i) for i in range(5)])})
 
     # add some nulls
-    df.ix[0, 's'] = None
-    df.ix[0, 'f'] = np.nan
-    df.ix[0, 'd'] = pd.NaT
+    df.iloc[0, df.columns.get_loc('s')] = None
+    df.iloc[0, df.columns.get_loc('f')] = np.nan
+    df.iloc[0, df.columns.get_loc('d')] = pd.NaT
+    df.iloc[0, df.columns.get_loc('o')] = np.array(np.nan)
 
     # JSON serialize. This will fail on non-sanitized dataframes
     df_clean = sanitize_dataframe(df)
@@ -46,6 +49,9 @@ def test_sanitize_dataframe():
 
     # Re-construct pandas dataframe
     df2 = pd.read_json(s)
+
+    # Re-order the columns to match df
+    df2 = df2[df.columns]
 
     # Re-apply original types
     for col in df:
@@ -56,7 +62,16 @@ def test_sanitize_dataframe():
         else:
             df2[col] = df2[col].astype(df[col].dtype)
 
+    # pandas doesn't properly recognize np.array(np.nan), so change it here
+    df.iloc[0, df.columns.get_loc('o')] = np.nan
     assert df.equals(df2)
+
+
+def test_sanitize_dataframe_infs():
+    df = pd.DataFrame({'x': [0, 1, 2, np.inf, -np.inf, np.nan]})
+    df_clean = sanitize_dataframe(df)
+    assert list(df_clean.dtypes) == [object]
+    assert list(df_clean['x']) == [0, 1, 2, None, None, None]
 
 
 def test_prepare_spec():

--- a/vega/utils.py
+++ b/vega/utils.py
@@ -28,8 +28,9 @@ def sanitize_dataframe(df):
     * Make a copy
     * Raise ValueError if it has a hierarchical index.
     * Convert categoricals to strings.
+    * Convert np.bool_ dtypes to Python bool objects
     * Convert np.int dtypes to Python int objects
-    * Convert floats to objects and replace NaNs by None.
+    * Convert floats to objects and replace NaNs/infs with None.
     * Convert DateTime dtypes into appropriate string representations
     """
     import pandas as pd
@@ -42,22 +43,38 @@ def sanitize_dataframe(df):
     if isinstance(df.columns, pd.core.index.MultiIndex):
         raise ValueError('Hierarchical indices not supported')
 
+    def to_list_if_array(val):
+        if isinstance(val, np.ndarray):
+            return val.tolist()
+        else:
+            return val
+
     for col_name, dtype in df.dtypes.iteritems():
         if str(dtype) == 'category':
             # XXXX: work around bug in to_json for categorical types
             # https://github.com/pydata/pandas/issues/10778
             df[col_name] = df[col_name].astype(str)
+        elif str(dtype) == 'bool':
+            # convert numpy bools to objects; np.bool is not JSON serializable
+            df[col_name] = df[col_name].astype(object)
         elif np.issubdtype(dtype, np.integer):
             # convert integers to objects; np.int is not JSON serializable
             df[col_name] = df[col_name].astype(object)
         elif np.issubdtype(dtype, np.floating):
-            # For floats, convert nan->None: np.float is not JSON serializable
-            col = df[col_name].astype(object)
-            df[col_name] = col.where(col.notnull(), None)
+            # For floats, convert to Python float: np.float is not JSON serializable
+            # Also convert NaN/inf values to null, as they are not JSON serializable
+            col = df[col_name]
+            bad_values = col.isnull() | np.isinf(col)
+            df[col_name] = col.astype(object).where(~bad_values, None)
         elif str(dtype).startswith('datetime'):
             # Convert datetimes to strings
             # astype(str) will choose the appropriate resolution
             df[col_name] = df[col_name].astype(str).replace('NaT', '')
+        elif dtype == object:
+            # Convert numpy arrays saved as objects to lists
+            # Arrays are not JSON serializable
+            col = df[col_name].apply(to_list_if_array, convert_dtype=False)
+            df[col_name] = col.where(col.notnull(), None)
     return df
 
 

